Tools/Algorithms/stencil.py

```diff
@@ -130,112 +133,127 @@ def compute_stencils(coeff_nodal, coeff_stagg, axis):
     coeff_stagg : numpy.ndarray
         Leading spectral staggered coefficient of the general PSATD equations.
     axis : int
-        Axis or direction (must be 0, 1, or 2).
+        Axis or direction (must be 0, 1, 2 or -1).
 
     Returns
     -------
-    (stencil_avg_nodal, stencil_avg_stagg) : tuple
-        Nodal and staggered stencils along a given direction.
+    stencils : dictionary
+        Dictionary of nodal and staggered stencils along a given direction.
     """
     # Inverse FFTs of the spectral coefficient along the chosen axis
     stencil_nodal = np.fft.ifft(coeff_nodal, axis=axis)
     stencil_stagg = np.fft.ifft(coeff_stagg, axis=axis)
 
     # Average results over remaining axes in spectral space
-    if axis == 0:
-          # Averaged over ky and kz
-          stencil_avg_nodal  = np.sum(np.sum(stencil_nodal, axis=2), axis=1)
-          stencil_avg_nodal /= (stencil_nodal.shape[2]*stencil_nodal.shape[1])
-          stencil_avg_stagg  = np.sum(np.sum(stencil_stagg, axis=2), axis=1)
-          stencil_avg_stagg /= (stencil_stagg.shape[2]*stencil_stagg.shape[1])
-    elif axis == 1:
-          # Averaged over kx and kz
-          stencil_avg_nodal  = np.sum(np.sum(stencil_nodal, axis=2), axis=0)
-          stencil_avg_nodal /= (stencil_nodal.shape[2]*stencil_nodal.shape[0])
-          stencil_avg_stagg  = np.sum(np.sum(stencil_stagg, axis=2), axis=0)
-          stencil_avg_stagg /= (stencil_stagg.shape[2]*stencil_stagg.shape[0])
-    elif axis == 2:
-          # Averaged over kx and ky
-          stencil_avg_nodal  = np.sum(np.sum(stencil_nodal, axis=1), axis=0)
-          stencil_avg_nodal /= (stencil_nodal.shape[1]*stencil_nodal.shape[0])
-          stencil_avg_stagg  = np.sum(np.sum(stencil_stagg, axis=1), axis=0)
-          stencil_avg_stagg /= (stencil_stagg.shape[1]*stencil_stagg.shape[0])
+    if dims == 1:
+        stencil_avg_nodal = stencil_nodal
+        stencil_avg_stagg = stencil_stagg
+    elif dims == 2:
+        if axis == 0:
+              # Averaged over kz
+              stencil_avg_nodal  = np.sum(stencil_nodal, axis=-1)
+              stencil_avg_nodal /= stencil_nodal.shape[-1]
+              stencil_avg_stagg  = np.sum(stencil_stagg, axis=-1)
+              stencil_avg_stagg /= stencil_stagg.shape[-1]
+        elif axis == 1 or axis == -1:
+              # Averaged over kx
+              stencil_avg_nodal  = np.sum(stencil_nodal, axis=0)
+              stencil_avg_nodal /= stencil_nodal.shape[0]
+              stencil_avg_stagg  = np.sum(stencil_stagg, axis=0)
+              stencil_avg_stagg /= stencil_stagg.shape[0]
+    elif dims == 3:
+        if axis == 0:
```

Comment:
Here's a trick I sometimes use that can remove these `if` blocks
```
# Average over the other two directions
i1 = (axis + 1) % 3
i2 = (axis + 2) % 3
stencil_avg_nodal  = (np.sum(np.sum(stencil_nodal, axis=i1), axis=i2)/
                                      (stencil_nodal.shape[i1]*stencil_nodal.shape[i2]))
stencil_avg_stagg  = (np.sum(np.sum(stencil_stagg, axis=i1), axis=i2)/
                                      (stencil_stagg.shape[i1]*stencil_stagg.shape[i2]))
```

Reply:
Here I'm thinking that maybe the `if` blocks make the code a bit easier to grasp, even though more verbose, so I would be inclinded to keep it as it is, if that's ok with you?

Reply:
Oh no, wait, I thought you were talking about the dimensional `if` blocks. I think I can implement the change you suggest here actually.

Reply:
Done in [2c21e08](https://github.com/ECP-WarpX/WarpX/pull/4016/commits/2c21e0848753de989ed1edbf5d502a62f7a8098f).